python/ray/data/datasource/file_based_datasource.py

```diff
@@ -134,11 +134,14 @@ def _get_write_path_for_block(
         *,
         filesystem: Optional["pyarrow.fs.FileSystem"] = None,
         dataset_uuid: Optional[str] = None,
-        block: Optional[ObjectRef[Block]] = None,
+        task_index: Optional[int] = None,
         block_index: Optional[int] = None,
         file_format: Optional[str] = None,
     ) -> str:
-        suffix = f"{dataset_uuid}_{block_index:06}.{file_format}"
+        if block_index is not None:
+            suffix = f"{dataset_uuid}_{task_index:06}_{block_index:06}.{file_format}"
+        else:
+            suffix = f"{dataset_uuid}_{task_index:06}.{file_format}"
```

Comment:
Maybe dumb question, but why do we encode information like `task_index` at all? Like, why don't we do something like `suffix = "{random_uuid}.{file_format}`?

Reply:
We have to do this to make sure each task writes to different files. That's also why I added the additional block_index in this PR. I'll add a comment.

Reply:
Wouldn't a random file name also guarantee that we write to different files?

Reply:
You don't want duplicate data on retries. Also, it's harder to debug with random IDs.